contracts/sol6/fprConversionRates/QtyStepConversionRates.sol

```diff
@@ -0,0 +1,443 @@
+            tokenData[token].baseSellRate = safeUint104(baseSell[ind]);
+        }
+
+        setCompactData(buy, sell, blockNumber, indices);
+    }
+
+    function setCompactData(
+        bytes14[] memory buy,
+        bytes14[] memory sell,
+        uint256 blockNumber,
+        uint256[] memory indices
+    ) public onlyOperator {
+        require(buy.length == sell.length, "buy-sell: mismatch length");
+        require(indices.length == buy.length, "buy-indices: mismatch length");
+        require(blockNumber <= 0xFFFFFFFF, "overflow block number");
+
+        for (uint256 i = 0; i < indices.length; i++) {
+            require(indices[i] < numCompactData, "invalid indices");
+            tokenRatesCompactData[indices[i]] = TokenRatesCompactData(
+                buy[i],
+                sell[i],
+                uint32(blockNumber)
+            );
+        }
+    }
+
+    function getRateUpdateBlock(IERC20Ext token) public view returns (uint256) {
+        uint256 compactDataArrayIndex = tokenData[token].compactDataArrayIndex;
+        return tokenRatesCompactData[compactDataArrayIndex].blockNumber;
+    }
+
+    function getRateWithoutImbalance(
+        TokenData memory data,
+        uint256 currentBlockNumber,
+        bool buy
+    ) internal view returns (uint256 rate, uint256 updateRateBlock) {
+        // get rate update block
+        TokenRatesCompactData memory compactData =
+            tokenRatesCompactData[data.compactDataArrayIndex];
+        updateRateBlock = compactData.blockNumber;
+        if (currentBlockNumber >= uint256(compactData.blockNumber) + validRateDurationInBlocks) {
+            return (0, 0); // rate is expired
+        }
+        rate = buy ? uint256(data.baseBuyRate) : uint256(data.baseSellRate);
+        // add rate update
+        uint256 fieldOffset = uint256(data.compactDataFieldIndex);
+        int8 rateUpdate =
+            buy ? int8(compactData.buy[fieldOffset]) : int8(compactData.sell[fieldOffset]);
+        rate = addBps(rate, int256(rateUpdate) * 10);
+    }
+
+    function getTokenQty(
+        IERC20Ext token,
+        uint256 ethQty,
+        uint256 rate
+    ) internal view returns (uint256) {
+        uint256 dstDecimals = getDecimals(token);
+        uint256 srcDecimals = ETH_DECIMALS;
+
+        return calcDstQty(ethQty, srcDecimals, dstDecimals, rate);
+    }
+
+    function executeStepFunction(StepFunctions storage f, int256 x)
+        internal
+        view
+        returns (int256)
+    {
+        uint256 len = f.length;
+        StepFunction memory step;
+        for (uint256 ind = 0; ind < len; ind++) {
+            step = f.data[ind];
+            if (x <= step.x) return int256(step.y);
+        }
+        return int256(step.y);
+    }
+
+    function addBps(uint256 rate, int256 bps) internal pure returns (uint256) {
+        require(rate <= MAX_RATE, "invalid rate");
+        require(bps >= MIN_BPS_ADJUSTMENT, "bps too low");
```

Comment:
bps can only be `extraBps` or `rateUpdate` in line 398
in 398, `rateUpdate` is converted to int8 so it should be a vali value.
We should only verify `extraBps` when it is set and remove redundant validation here

Reply:
This is good for gas consumption.

But for safeMath protection, range check is necessary to make sure the function is safe.

Reply:
but it is internal function and we know the exact places that call this function right?

Reply:
Anyway, I will add the range check when setting step function to avoid revert

Reply:
> but it is internal function and we know the exact places that call this function right?

Yes, but it will much more harder to review safe-math (complex range assumption). Meanwhile, removing 2 condition - only reduce 100 gas for query rate.
getConversionRate gas used: 25848 stepIndex=0 -> 25744
getConversionRate gas used: 35670 stepIndex=9 -> 35566

Do you still want to remove this